### PR TITLE
Improvments

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

**Comment by MaxGraey** [comment]
Alright. I refactored and get rig of all float point operations from original implementation. Also pin base size to 58 which simplify many things but keep room for customize values of alphabet.

File: assembly/index.ts [hunk 1]
```diff
@@ -26,110 +40,93 @@ function FACTOR(length: i32): i32 {
...
+    for (let it = size - 1; it != -1 && (carry != 0 || i < length); --it, ++i) {
+      carry += u32(b58[it]) << 8;
+      b58[it] = carry % BASE;
+      carry = carry / BASE;
     }
-
-    assert(carry == 0, 'Non-zero carry');
-    length = i
-    pbegin++
+    if (ASC_OPTIMIZE_LEVEL == 0) {
```
Comment (willemneal) [inline assembly/index.ts:63]: Why this branch?

**Review by willemneal — APPROVED** [review]
LGTM, just want one answer.

File: assembly/index.ts (hunk 1 above)
Comment (MaxGraey) [reply assembly/index.ts:63]: Could `Non-zero carry` happened with fixed alphabet set? I mean is `assert(!carry, 'Non-zero carry')` meaningful runtime check or just debug verification?

Comment (MaxGraey) [reply assembly/index.ts:63]: up

Comment (willemneal) [reply assembly/index.ts:63]: I think it's a debug check.  I think this is good to go.

**Pull request merged**